### PR TITLE
add mixer

### LINKED ISSUE #31: Liquid Mixer UI
Screenshot_20200515-123543
Blocked by #30

### PR DESCRIPTION
**Pull request by gbaranski**
Need to add proper fetching remote api server, my proposal is modal with results. Fix responsiveness and change icon in config/routes, remove redux?. Now it looks like that

Screenshot 2020-05-29 at 14 47 24

closes #31

### CODE REVIEW

File: src/pages/mixer/Mixer.tsx [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import { connect } from 'dva';
+import ButtonGroup from 'antd/es/button/button-group';
+import { Dispatch } from 'redux';
+import { ConnectState } from '@/models/connect';
+import { FormattedMessage } from 'umi-plugin-react/locale';
+import { UpOutlined } from '@ant-design/icons';
+import InputElements from './inputElements';
+import SelectType from './SelectType';
+
+export interface OhmLawProps {
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:13]: MixerProps

File: src/pages/mixer/Mixer.tsx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  ohm: OhmModelState;
+  dispatch: Dispatch;
+}
+
+// TODO check if not needed adjust to new Form API
+const OhmLaw: React.FC<OhmLawProps> = () => {
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:19]: `const Mixer`

File: src/pages/mixer/Mixer.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  const mixDataPattern = {
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:20]: Use `Mixable` type from `@vapetool/types`

File: src/pages/mixer/Mixer.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  };
+
+  const [mixData1, setMixData1] = useState(mixDataPattern);
+
+  const [mixData2, setMixData2] = useState(mixDataPattern);
+
+  const handleClear = () => {
+    setMixData1({
+      ...mixDataPattern,
+    });
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:34]: `setMixData2` as well

File: src/pages/mixer/Mixer.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  };
+
+  //   const [result, setResult] = useState({
+  //     totalMl: 0,
+  //     pgRatio: 50,
+  //     strength: 0,
+  //   });
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:41]: remove unused comment

File: src/pages/mixer/inputElements.tsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
+import React from 'react';
+import { InputNumber } from 'antd';
+import { formatMessage } from 'umi-plugin-react/locale';
+import FormItem from 'antd/lib/form/FormItem';
+import VgPgRatioView from '@/components/VgPgRatioView';
+
+export default function InputElements(props: any) {
+  return (
+    <>
+      Amount [ml]
```
Comment (stasbar) [inline src/pages/mixer/inputElements.tsx:10]: Use label element, like this in `OhmLaw` page

File: src/pages/mixer/inputElements.tsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+              ml: newValue,
+            });
+          }}
+          placeholder={formatMessage({
+            id: 'misc.units.long.ml',
+            defaultMessage: 'Amount [ml]',
+          })}
+        />
+      </FormItem>
+      Strength [mg/ml]
```
Comment (stasbar) [inline src/pages/mixer/inputElements.tsx:31]: Same

File: src/pages/mixer/inputElements.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+          size="large"
+          step={1}
+          min={0.5}
+          precision={2}
+          style={{ width: '100%', maxWidth: 200 }}
+          onChange={newValue => {
+            props.onValueChange({
+              ...props.mixData,
+              ml: newValue,
+            });
```
Comment (stasbar) [inline src/pages/mixer/inputElements.tsx:23]: Good job, but change `ml` to `amount`

File: src/pages/mixer/inputElements.tsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+          value={props.mixData.mg_ml}
+          size="large"
+          step={0.5}
+          min={0}
+          precision={2}
+          style={{ width: '100%', maxWidth: 200 }}
+          onChange={newValue => {
+            props.onValueChange({
+              ...props.mixData,
+              mg_ml: newValue,
```
Comment (stasbar) [inline src/pages/mixer/inputElements.tsx:43]: change to `strength`

File: src/pages/mixer/inputElements.tsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            id: 'misc.units.long.nicotine',
+            defaultMessage: 'Strength [mg/ml]',
+          })}
+        />
+      </FormItem>
+      <VgPgRatioView
+        onRatioChange={newValue => {
+          props.onValueChange({
+            ...props.mixData,
+            vgRatio: 100 - newValue,
```
Comment (stasbar) [inline src/pages/mixer/inputElements.tsx:56]: `vgRatio` to `ratio`

File: src/pages/mixer/Mixer.tsx [hunk 11]
```diff
@@ -18,34 +19,40 @@ export interface OhmLawProps {
...
   const handleCalculate = async (e: any) => {
     e.preventDefault();
+    console.log(mixable1, mixable2);
+    const headers = new Headers();
+    headers.append('mixable1', JSON.stringify(mixable1));
+    headers.append('mixable2', JSON.stringify(mixable2));
+    fetch('url', {
+      method: 'POST',
+      headers,
+    })
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:51]: Replace with
```
    fetch('url', {
      method: 'POST',
body: JSON.stringify({ mixable1, mixable2 })
    })
```

File: src/pages/mixer/tableData.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    key: 'name',
+  },
+  {
+    title: 'Percentage [%]',
+    dataIndex: 'percentage',
+    key: 'percentage',
+  },
+  {
+    title: 'Amount [ml]',
+    dataIndex: 'amount',
```
Comment (stasbar) [inline src/pages/mixer/tableData.ts:14]: change to `ml`

File: src/pages/mixer/tableData.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    key: 'amount',
```
Comment (stasbar) [inline src/pages/mixer/tableData.ts:15]: change to `ml`

File: src/pages/mixer/tableData.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+    ellipsis: {
+      showTitle: false,
+    },
+  },
+  {
+    title: 'Drops',
+    dataIndex: 'drops',
```
Comment (stasbar) [inline src/pages/mixer/tableData.ts:22]: change to `drips`

File: src/pages/mixer/tableData.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    key: 'drops',
```
Comment (stasbar) [inline src/pages/mixer/tableData.ts:23]: change to `drips`

File: src/pages/mixer/Mixer.tsx [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+    strength: 0,
+  };
+
+  const [mixable1, setMixable1] = useState(mixDataPattern);
+
+  const [mixable2, setMixable2] = useState(mixDataPattern);
+
+  const [results, setResults] = useState({
+    ...mixResultPattern,
+    results: [mixResultPattern, mixResultPattern, mixResultPattern],
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:44]: remove this line

File: src/pages/mixer/Mixer.tsx [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+    },
+    {
+      key: '3',
+      name: 'Total',
+      percentage: 100,
+      amount: results.results[1].ml + results.results[2].ml,
+      drops: results.results[1].drips + results.results[2].drips,
+      weight: results.results[1].weight + results.results[2].weight,
+    },
+  ];
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:102]: remove

File: src/pages/mixer/Mixer.tsx [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+                    <Col style={{ margin: 'auto' }}>
+                      <Row>
+                        <Typography style={{ fontWeight: 'bold', fontSize: 24 }}>
+                          Strength:&nbsp;
+                        </Typography>
+                        <Typography>{results.strength}mg/ml</Typography>
+                      </Row>
+                    </Col>
+                  </Row>
+                  <Table columns={columns} dataSource={data} pagination={false} />
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:160]: change to `dataSource={results.results}`

File: src/pages/mixer/Mixer.tsx [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+    strength: undefined,
+    ratio: 50,
+    thinner: undefined,
+    flavorPercentage: undefined,
+  };
+
+  const mixResultPattern: Partial<MixResult> = {
+    amount: 0,
+    ratio: 0,
+    strength: 0,
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:35]: add 
`results: [],`

File: src/pages/mixer/Mixer.tsx [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+  const mixDataPattern: Partial<Mixable> = {
+    type: MixableType.BASE,
+    amount: undefined,
+    strength: undefined,
+    ratio: 50,
+    thinner: undefined,
+    flavorPercentage: undefined,
+  };
+
+  const mixResultPattern: Partial<MixResult> = {
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:32]: no need to `Partial<MixResult>`, just `MixResult` if you specify `results: [],`

**Comment by gbaranski** [comment]
@stasbar please review

File: src/pages/mixer/Mixer.tsx [hunk 21]
```diff
@@ -10,27 +10,18 @@ import SelectType from './SelectType';
...
     strength: undefined,
     ratio: 50,
     thinner: undefined,
-    flavorPercentage: undefined,
   };
 
   const mixResultPattern: Partial<MixResult> = {
-    amount: 0,
+    // @ts-ignore
+    results: [mixDataPattern, mixDataPattern],
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:24]: There should be no results before the first calculation, so you can place empty `results: [],`

File: src/pages/mixer/Mixer.tsx [hunk 22]
```diff
@@ -145,15 +136,20 @@ const Mixer: React.FC = () => {
                         <Typography style={{ fontWeight: 'bold', fontSize: 24 }}>
                           Ratio:&nbsp;
                         </Typography>
-                        <Typography>55VG/45PG</Typography>
+                        <Typography>
+                          {Math.round(100 - results.ratio * 10) / 10}VG/
+                          {Math.round(results.ratio * 10) / 10}PG
```
Comment (stasbar) [inline src/pages/mixer/Mixer.tsx:141]: No decimal places here, VG/PG are integers, so better do 
```
{100 - Math.round(results.ratio)}VG
{Math.round(results.ratio)}PG
```

File: src/services/mixer.ts [hunk 23]
```diff
@@ -2,14 +2,11 @@ import { Mixable } from '@vapetool/types';
...
-export function calculate(mixable1: Partial<Mixable>, mixable2: Partial<Mixable>) {
+export function calculate(mixable1: Mixable, mixable2: Mixable) {
   return sendRequest(mixable1, mixable2);
 }
 
-export async function sendRequest<T>(
-  mixable1: Partial<Mixable>,
-  mixable2: Partial<Mixable>,
-): Promise<T> {
+export async function sendRequest<T>(mixable1: Mixable, mixable2: Mixable): Promise<T> {
```
Comment (stasbar) [inline src/services/mixer.ts]: You can merge those functions, also the generic type `<T>` is unnecessary
so merge those two functions to :
```ts
export async function calculate(mixable1: Mixable, mixable2: Mixable): Promise<MixResult> {
  try {
    if (!auth.currentUser) {
      throw Error('You are not logged in');
    }
    const idToken = await auth.currentUser!.getIdToken(false);
    const response =  await request.post(`/api/calculator/mixer/mix`, {
      data: { mixable1, mixable2, idToken },
    });
    return response.json()
  } catch (e) {
    console.error(e);
    throw e;
  }
}
```

**Pull request merged**